api/shared/condition_types.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2020 Red Hat
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// +kubebuilder:object:generate:=true
+
+package shared
+
+import (
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Conditions for status in web console
+
+// ConditionDetails used for passing condition information into generic functions
+// e.f. GetDataFromSecret
+type ConditionDetails struct {
+	ConditionNotFoundType   ConditionType
+	ConditionNotFoundReason ConditionReason
+	ConditionErrorType      ConditionType
+	ConditionErrordReason   ConditionReason
+}
+
+// ConditionList - A list of conditions
+type ConditionList []Condition
+
+// Condition - A particular overall condition of a certain resource
+type Condition struct {
+	Type               ConditionType          `json:"type"`
+	Status             corev1.ConditionStatus `json:"status"`
+	Reason             ConditionReason        `json:"reason,omitempty"`
+	Message            string                 `json:"message,omitempty"`
+	LastHeartbeatTime  metav1.Time            `json:"lastHearbeatTime,omitempty"`
+	LastTransitionTime metav1.Time            `json:"lastTransitionTime,omitempty"`
+}
+
+// ConditionType - A summarizing name for a given condition
+type ConditionType string
+
+// ConditionReason - Why a particular condition is true, false or unknown
+type ConditionReason string
+
+// NewCondition - Create a new condition object
+func NewCondition(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) Condition {
+	now := metav1.Time{Time: time.Now()}
+	condition := Condition{
+		Type:               conditionType,
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		LastHeartbeatTime:  now,
+		LastTransitionTime: now,
+	}
+	return condition
+}
+
+// Set - Set a particular condition in a given condition list
+func (conditions *ConditionList) Set(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) {
+	condition := conditions.Find(conditionType)
+
+	// If there isn't condition we want to change, add new one
```

Comment:
can it hold multiple conditions of the same Type? Find would return the first found then.
If it can't, maybe use Type as a dict in the Conditions, instead of a list?

Reply:
I think I got it. Set() finds the "previous" condition of the same type. So the head is the current and the previous one. So a map does not fit. tricky.

Reply:
yes, it just sets/adds/updates the latest condition in the condition list